WalletWasabi/Microservices/ProcessBridge.cs

```diff
@@ -16,79 +15,109 @@ public class ProcessBridge : IProcessBridge
 		public ProcessBridge(string processPath)
 		{
 			ProcessPath = Guard.NotNullOrEmptyOrWhitespace(nameof(processPath), processPath);
+
 			if (!File.Exists(ProcessPath))
 			{
-				var fileNameWithoutExtension = Path.GetFileNameWithoutExtension(ProcessPath);
-				throw new FileNotFoundException($"{fileNameWithoutExtension} is not found.", ProcessPath);
+				throw new FileNotFoundException($"{Path.GetFileNameWithoutExtension(ProcessPath)} is not found.", ProcessPath);
 			}
 		}
 
 		public string ProcessPath { get; }
 
 		public Process Start(string arguments, bool openConsole)
 		{
-			var finalArguments = arguments;
-			var redirectStandardOutput = !openConsole;
-			var useShellExecute = openConsole;
-			var createNoWindow = !openConsole;
-			var windowStyle = openConsole ? ProcessWindowStyle.Normal : ProcessWindowStyle.Hidden;
+			var process = CreateProcessInstance(arguments, openConsole);
+			StartProcess(process);
+			return process;
+		}
 
-			if (openConsole && !RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+		public async Task<(string response, int exitCode)> SendCommandAsync(string arguments, bool openConsole, CancellationToken cancel)
+		{
+			var taskCompletionSource = new TaskCompletionSource<bool>();
+
+			// "using" makes sure the process exits at the end of this method.
+			using var process = CreateProcessInstance(arguments, openConsole);
+			process.EnableRaisingEvents = true;
+			process.Exited += (object sender, EventArgs e) => taskCompletionSource.TrySetResult(true);
+
+			StartProcess(process);
+
+			try
 			{
-				throw new PlatformNotSupportedException($"{RuntimeInformation.OSDescription} is not supported.");
-				//var escapedArguments = (hwiPath + " " + arguments).Replace("\"", "\\\"");
-				//useShellExecute = false;
-				//if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
-				//{
-				//	fileName = "xterm";
-				//	finalArguments = $"-e \"{escapedArguments}\"";
-				//}
-				//else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
-				//{
-				//	fileName = "osascript";
-				//	finalArguments = $"-e 'tell application \"Terminal\" to do script \"{escapedArguments}\"'";
-				//}
+				// If this token is already in the canceled state, the delegate will be run immediately and synchronously.
+				using (cancel.Register(() => taskCompletionSource.TrySetCanceled()))
+				{
+					await taskCompletionSource.Task;
+				}
 			}
-
-			ProcessStartInfo startInfo = new ProcessStartInfo
+			catch (OperationCanceledException ex)
 			{
-				FileName = ProcessPath,
-				Arguments = finalArguments,
-				RedirectStandardOutput = redirectStandardOutput,
-				UseShellExecute = useShellExecute,
-				CreateNoWindow = createNoWindow,
-				WindowStyle = windowStyle
-			};
+				try
+				{
+					process.Kill(entireProcessTree: true);
```

Comment:
I disagree with killing process in this way. This really kills a process immediately without allowing it to finish in a clean way. I personally have suffered files corruptions in Bitcoin Core by killing its process.

Reply:
Right. That's quite forceful solution. I have update the PR.